fix: Not able to create consecutive ooo

vijayraghav-io opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #17387 
- Fixes CAL-XXXX 

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

**Issue:**
Consecutive OOO Issue:
https://www.loom.com/share/3bb5b7022156496b8ab936bb5e161cbf?sid=a5392921-b3a6-4e47-80f2-4951e80aa811

Consecutive Reverse Redirect Issue:
https://www.loom.com/share/f1955fe640b24952adc378c58073b5aa?sid=d28615ac-9d1c-4ca8-b85b-097041441b34

**Fix:**
https://www.loom.com/share/aaa9fd01fcb440738ffa6e144e4bb59e?sid=87f987e5-0e29-4719-9845-0f125a9c5c2c

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] - N/A - I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
fix: able to create consecutive ooo

AliPythonDev commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -32,6 +32,8 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
 
   const inputStartTime = dayjs(startDate).startOf("day");
   const inputEndTime = dayjs(endDate).endOf("day");
+  const startDateUtc = dayjs.utc(startDate).add(input.offset, "minute");
+  const endDateUtc = dayjs.utc(endDate).add(input.offset, "minute");
```
Good use of startDateUtc and endDateUtc for consistent date handling. Should help avoid time zone issues

AlleyNawaz reviewed (dismissed):
Looks good overall! 
Just a couple of small tweaks noted, nice work on improving the OOO checks.

keithwillcode commented:
> Looks good overall! Just a couple of small tweaks noted, nice work on improving the OOO checks.

@AliPythonDev Has this been tested to ensure it works? I see no automated tests were modified in this PR (which should happen)

vijayraghav-io commented:
> > Looks good overall! Just a couple of small tweaks noted, nice work on improving the OOO checks.
> 
> @AliPythonDev Has this been tested to ensure it works? I see no automated tests were modified in this PR (which should happen)

yes sure will add/update tests.

keithwillcode reviewed (dismissed):
Marking as requested changes while @vijayraghav-io updates the tests.

vijayraghav-io committed:
added e2e test

vijayraghav-io committed:
chore

vijayraghav-io commented:
@keithwillcode , added an e2e test.

PFB the snapshot of same e2e when applied on the current main branch, just to show a false positive / getting error as expected when 2nd consecutive ooo is created on current main branch.

<img width="1108" alt="Screenshot 2024-11-01 at 3 28 21 PM" src="https://github.com/user-attachments/assets/cbd6309b-0414-4e2e-b104-94fbdbfaa255">

Praashh reviewed (changes requested):
Hey @vijayraghav-io, if we allow overlapping OOO, then there is no point in checking if the OOO entry for dates already exists. So, in `packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts` we can simply get remove of outOfOfficeEntry.

vijayraghav-io committed:
update to allow overlapping ooo and test for same

vijayraghav-io commented:
@Praashh , updated to allow overlapping OOOs and also added test for same

vijayraghav-io commented:
All tests are passing.

Praashh reviewed (commented):
hey @vijayraghav-io can you tell me where are you handling date null condition ?

vijayraghav-io commented:
> hey @vijayraghav-io can you tell me where are you handling date null condition ?

it's not clear what is meant by handling date null condition here,  currently we can't create ooo without entering a date or picking a date, so i don't think it's required to handle date null condition explicitly at backend.

Praashh commented:
> > hey @vijayraghav-io can you tell me where are you handling date null condition ?
> 
> it's not clear what is meant by handling date null condition here, currently we can't create ooo without entering a date or picking a date, so i don't think it's required to handle date null condition explicitly at backend.

there would be minDate option that sets by default to new Date()  and nullable to allow all dates enabled.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_createConsecutiveOOO

vijayraghav-io committed:
refactor to reduce repeating code in tests

vijayraghav-io committed:
nit

vijayraghav-io commented:
> there would be minDate option that sets by default to new Date() and nullable to allow all dates enabled.

The scope of this PR is to fix with regards to correct place to do the utc conversion of input dates and move it prior to doing other input date validations. Because of this issue , consecutive OOOs were not possible to create. Also might have caused other unforeseen issues w.r.t timezones.
Knowing that there is another issue created to allow overlapping OOOs, so as requested this requirement was also handled in this PR, also test case for allowing overlapping  is added.

But, handling minDate might be for other independent requirement and may not be in scope for this PR or issue being handled IMO.
Please let me know or confirm if any additional req or case need to be covered here, i can include the same and also add tests for it.

vijayraghav-io committed:
added testcases for consecutive and reverse redirects

vijayraghav-io commented:
> Marking as requested changes while @vijayraghav-io updates the tests.

Hi @keithwillcode , have added all tests.

Praashh commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
it's not handling duplicate OOOs.
<img width="931" alt="Screenshot 2024-11-19 at 5 14 00 PM" src="https://github.com/user-attachments/assets/69acaa48-fd44-4923-bcd1-80ef200e1155">

Praashh replied:
fixed

Praashh commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -73,51 +75,6 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
-        {
-          uuid: {
-            not: input.uuid ?? "",
-          },
-        },
-        {
-          OR: [
-            {
-              start: {
-                lt: inputEndTime.toISOString(), //existing start is less than or equal to input end time
-              },
-              end: {
-                gt: inputStartTime.toISOString(), //existing end is greater than or equal to input start time
-              },
-            },
-            {
-              //existing start is within the new input range
-              start: {
-                gt: inputStartTime.toISOString(),
-                lt: inputEndTime.toISOString(),
-              },
-            },
-            {
-              //existing end is within the new input range
-              end: {
-                gt: inputStartTime.toISOString(),
-                lt: inputEndTime.toISOString(),
-              },
-            },
-          ],
-        },
-      ],
-    },
-  });
-
-  // don't allow overlapping entries
-  if (outOfOfficeEntry) {
-    throw new TRPCError({ code: "CONFLICT", message: "out_of_office_entry_already_exists" });
-  }
-
```
```// Validate if OOO entry for these dates already exists
  const outOfOfficeEntry = await prisma.outOfOfficeEntry.findFirst({
    where: {
      AND: [
        { userId: ctx.user.id },
        {
          uuid: {
            not: input.uuid ?? "",
          },
        },
        {
          OR: [
            {
              start: { lte: endDateUtc.toDate() },
              end: { gte: startDateUtc.toDate() },
            },
          ],
        },
      ],
    },
  });

  // don't allow overlapping entries
  if (outOfOfficeEntry) {
    throw new TRPCError({ code: "CONFLICT", message: "out_of_office_entry_already_exists" });
  }```
  
  this fixs, please add this

Praashh committed:
fix:duplicate OOO creation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=17388) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=17388) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: vijayraghav-io<br/>:x: Praashh<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=17388) it.</sub>

Praashh commented:
previous issue :
![image](https://github.com/user-attachments/assets/396ed935-7328-4b86-af27-f529f2248d3f)

fix:

https://github.com/user-attachments/assets/cc4ea603-ad07-4e37-8991-06fb5e78f97b

yeah now fixed the duplicate OOO.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_createConsecutiveOOO

vijayraghav-io committed:
prevent duplicate entries, also added test

vijayraghav-io committed:
chore

Praashh commented:
> @Praashh , updated to restrict duplicate ooo entries, also added test for the same.

I have solved it though, but thank you for the test cases.

vijayraghav-io commented:
> > @Praashh , updated to restrict duplicate ooo entries, also added test for the same.
> 
> I have solved it though, but thank you for the test cases.

An update in where condition was required, It has to be 

<img width="735" alt="image" src="https://github.com/user-attachments/assets/1e219419-7253-404e-a620-874f8e797cdf">

Not 
<img width="441" alt="image" src="https://github.com/user-attachments/assets/6da31bf6-27f7-45d1-9b47-337d51870f24">

Also, one of the e2e test case was failing bcoz of this, hence updated.

Praashh commented:
> > > @Praashh , updated to restrict duplicate ooo entries, also added test for the same.
> > 
> > 
> > I have solved it though, but thank you for the test cases.
> 
> An update in where condition was required, It has to be
> 
> <img alt="image" width="735" src="https://private-user-images.githubusercontent.com/3139634/387782030-1e219419-7253-404e-a620-874f8e797cdf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNDAxMDksIm5iZiI6MTczMjAzOTgwOSwicGF0aCI6Ii8zMTM5NjM0LzM4Nzc4MjAzMC0xZTIxOTQxOS03MjUzLTQwNGUtYTYyMC04NzRmOGU3OTdjZGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTgxMDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODQ0NTUzZGJmZmE5MTA4Y2UwZjgzMzVmMjI1NTI1NGM5NWUzZGY0ZTYzN2QyYTY5NDA2MzdiMjUyYTk5Yjk0NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.eO4zUiBLZ5xAveg_hFiAUPuI5Nfvmrm3TMu47l68BKE">
> Not <img alt="image" width="441" src="https://private-user-images.githubusercontent.com/3139634/387783603-6da31bf6-27f7-45d1-9b47-337d51870f24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNDAxMDksIm5iZiI6MTczMjAzOTgwOSwicGF0aCI6Ii8zMTM5NjM0LzM4Nzc4MzYwMy02ZGEzMWJmNi0yN2Y3LTQ1ZDEtOWI0Ny0zMzdkNTE4NzBmMjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTgxMDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTBkNTcwZjYyZTU5MzJlOTM0OGM4MjNlYjY1YTVjOGUxNzZhYWFjMTE3Nzk2MWJhZjA2MDM4YzM4MjE3NTU5OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.v_AKCvhCjyrTF_B3UwReLG-7c-HTlOVAGgZEXytbFfg">
> 
> Also, one of the e2e test case was failing bcoz of this, hence updated.

alrighty, thanks a lot.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_createConsecutiveOOO

vijayraghav-io committed:
update after merges

vijayraghav-io committed:
correct datetype in prisma query

anikdhabal commented:
@vijayraghav-io could you pls fix the conflicts🙏

vijayraghav-io commented:
> @vijayraghav-io could you pls fix the conflicts🙏

sure will fix 🙏

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into fix_createConsecutiveOOO

vijayraghav-io committed:
update to changes in rdp after merge

Praashh reviewed (approved):
All good, LGTM!!

vijayraghav-io commented:
Just for information, this PR already had same fix done from https://github.com/calcom/cal.com/pull/17756/files

Anyway retained the new testcases for consecutive ooo check and removed the similar test cases that was also added by other PR after merge

vijayraghav-io commented:
@anikdhabal , resolved conflicts
